src/Functions/generateUUIDv7.cpp

```diff
@@ -49,76 +50,93 @@ uint64_t getTimestampMillisecond()
+/// 3. calculate the capacity of the range, which could be lower than `input_rows_count`
+UUIDv7Range getRangeOfAvailableParts(const UUIDv7 & available, size_t input_rows_count)
+{
+    /// 1. `now`
+    UUIDv7 begin = {.timestamp = getTimestampMillisecond(), .counter = getRandomCounter()};
+
+    /// 2. `begin`
+    bool available_is_now_or_later = begin.timestamp <= available.timestamp;
+    bool available_is_too_far_ahead = begin.timestamp + 10000 <= available.timestamp;
+    if (available_is_now_or_later && !available_is_too_far_ahead)
     {
-        const uint64_t counter_low_bits = (UUIDHelpers::getLowBytes(uuid) >> rand_b_low_bits_count) & counter_low_bits_mask;
-        const uint64_t counter_high_bits = UUIDHelpers::getHighBytes(uuid) & counter_high_bits_mask;
-        counter = (counter_high_bits << 30) | counter_low_bits;
+        begin.timestamp = available.timestamp;
+        begin.counter = available.counter;
     }
 
-    void incrementCounter(UUID & uuid)
+    /// 3. `capacity`
+    UUIDv7 end;
+    const uint64_t counter_nums_in_current_timestamp_left = (counter_limit - begin.counter);
+    uint64_t capacity = input_rows_count;
+
+    if (input_rows_count >= counter_nums_in_current_timestamp_left) [[unlikely]]
     {
-        if (++counter == counter_limit) [[unlikely]]
-        {
-            ++last_timestamp;
-            resetCounter(uuid);
-            setTimestampAndVersion(uuid, last_timestamp);
-            setVariant(uuid);
-        }
-        else
-        {
-            UUIDHelpers::getHighBytes(uuid) = (last_timestamp << 16) | 0x7000 | (counter >> counter_low_bits_count);
-            UUIDHelpers::getLowBytes(uuid) = (UUIDHelpers::getLowBytes(uuid) & rand_b_with_counter_bits_mask) | variant_2_mask | ((counter & counter_low_bits_mask) << rand_b_low_bits_count);
-        }
+        capacity = counter_nums_in_current_timestamp_left;
+        /// The next range has to start on the next timestamp, with a random counter
+        end.timestamp = begin.timestamp + 1;
+        end.counter = getRandomCounter();
     }
-
-    void generate(UUID & uuid, uint64_t timestamp)
+    else
     {
-        const bool need_to_increment_counter = (last_timestamp == timestamp) || ((last_timestamp > timestamp) & (last_timestamp < timestamp + 10000));
-        if (need_to_increment_counter)
-        {
-            incrementCounter(uuid);
-        }
-        else
-        {
-            last_timestamp = timestamp;
-            resetCounter(uuid);
-            setTimestampAndVersion(uuid, last_timestamp);
-            setVariant(uuid);
-        }
+        end.timestamp = begin.timestamp;
+        end.counter = begin.counter + capacity;
     }
-};
+
+    return {begin, end, capacity};
+}
 
 
 struct Data
 {
     /// Guarantee counter monotonicity within one timestamp across all threads generating UUIDv7 simultaneously.
-    static inline CounterFields fields;
+    static inline UUIDv7 lowest_available_parts;
```

Comment:
It would be great to use CAS and an atomic "lowest available UUID" variable here, similar to the implementation in generateSnowflakeId.

Reply:
We have 48 + 42 bits of data, so I'm not sure if it is possible. At least, simply using `std::atomic<UUIDv7>`  doesn't work. Something else is needed.